### PR TITLE
Make builder pattern more practical

### PR DESCRIPTION
**Pull request by TGotwig**
https://github.com/TGotwig/patterns/blob/patch-1/patterns/builder.md

### CODE REVIEW

**Comment by nesteruk** [comment]
It is confusing to have `bar: String` in both `Foo` and `FooBuilder` — consider renaming one of the `bar`s to something else.

**Comment by TGotwig** [comment]
@nesteruk Changed this and a little bit more 🙂

File: patterns/builder.md [hunk 1]
```diff
@@ -8,38 +8,53 @@ Construct an object with calls to a builder helper.
...
+    bar: String,
 }
 
-struct FooBuilder {
+// Optional, define the textual representation for Foo.
+impl fmt::Display for Foo {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "bar: {}", self.bar)
+    }
+}
```
Comment (lambda-fairy) [inline patterns/builder.md:23]: This `Display` impl feels irrelevant to illustrating the builder pattern. Can we replace this with `#[derive(Debug)]`, and use `format!("{:?}", foo)` in the test instead?

**Review by lambda-fairy — DISMISSED** [review]
LGTM, modulo the comment below

File: patterns/builder.md (hunk 1 above)
Comment (TGotwig) [reply patterns/builder.md:23]: I will look for this soon 🙂

**Comment by pickfire** [comment]
This example does not seemed to be explaining much, maybe we could have a real-life example or an explanation on why it needs to be constructed this way rather than just specifying the struct directly?

File: patterns/builder.md (hunk 1 above)
Comment (TGotwig) [reply patterns/builder.md:14]: Yeah you are right! Changed this up 👷

**Comment by TGotwig** [comment]
@pickfire Hmm, changed it a little bit up.
I think that runnable code is better than not runnable code and the comments should explain enough 🧐

File: patterns/builder.md [hunk 2]
```diff
@@ -2,46 +2,51 @@
...
+
+  pub fn name(mut self, bar: &str) -> Builder {
+    // Set the name on the builder itself, and return the builder by value.
+    self.builder_bar = bar.to_string();
+    self
+  }
+
+  // If we can get away with not consuming the Builder here, that is an
+  // advantage. It means we can use the builder as a template for constructing many Foos.
+  pub fn finish(self) -> Foo {
```
Comment (pickfire) [inline patterns/builder.md:37]: Isn't it usually `build`?

**Review by pickfire — COMMENTED** [review]
The indentation seemed off, should be 4 spaces rather than 2 spaces.

File: patterns/builder.md [hunk 3]
```diff
@@ -2,46 +2,51 @@
...
+  // Lots of complicated fields.
+  bar: String,
 }
 
-struct FooBuilder {
-    // Probably lots of optional fields.
-    ...
+pub struct Builder {
+  // Probably lots of optional fields.
+  builder_bar: String,
```
Comment (pickfire) [inline patterns/builder.md:18]: We can probably just keep the same `bar` here.

File: patterns/builder.md [hunk 4]
```diff
@@ -2,46 +2,49 @@
...
 ```rust
-struct Foo {
+#[derive(Debug)]
+pub struct Foo {
     // Lots of complicated fields.
+    bar: String,
 }
 
-struct FooBuilder {
+pub struct Builder {
```
Comment (pickfire) [inline patterns/builder.md:16]: Why shouldn't we keep the builder name as `FooBuilder`?

File: patterns/builder.md [hunk 5]
```diff
@@ -2,46 +2,49 @@
...
-    // various aspects of a Foo.
-    ...
-
     // If we can get away with not consuming the Builder here, that is an
-    // advantage. It means we can use the builder as a template for constructing
-    // many Foos.
-    fn finish(&self) -> Foo {
-        // Create a Foo from the FooBuilder, applying all settings in FooBuilder to Foo.
+    // advantage. It means we can use the builder as a template for constructing many Foos.
+    pub fn finish(self) -> Foo {
```
Comment (pickfire) [inline patterns/builder.md:37]: Some parts seemed inconsistent, some uses `finish` and below uses `build`.

Comment (TGotwig) [reply patterns/builder.md]: Unfortunately I had this overwritten, but now it is `build` again 👷

File: patterns/builder.md (hunk 4 above)
Comment (TGotwig) [reply patterns/builder.md]: On second thought you are right, changed it back ↩️

File: patterns/builder.md [hunk 6]
```diff
@@ -2,46 +2,51 @@
...
     }
+  }
 }
 
-fn main() {
-    let f = FooBuilder::new().named("Bar").with_attribute(...).finish();
+#[test]
+fn builder_test() {
+  let foo: Foo = Builder::new().name("Y").finish();
+  assert_eq!(format!("{:?}", foo), "Foo { bar: \"Y\" }");
```
[the remark this answers is not in the document]
Comment (TGotwig) [reply patterns/builder.md]: Changed this up 👷

File: patterns/builder.md [hunk 7]
```diff
@@ -2,46 +2,58 @@
...
+pub struct Foo {
     // Lots of complicated fields.
+    bar: String,
+}
+
+impl PartialEq for Foo {
+    fn eq(&self, other: &Self) -> bool {
+        self.bar == other.bar
+    }
 }
```
Comment (pickfire) [inline patterns/builder.md:20]: Why not just do derive `PartialEq` above? Rather than writing this from scratch?

File: patterns/builder.md [hunk 8]
```diff
@@ -2,46 +2,58 @@
...
 
-fn main() {
-    let f = FooBuilder::new().named("Bar").with_attribute(...).finish();
+#[test]
+fn builder_test() {
+    let foo = Foo {
+        bar: String::from("Y"),
+    };
+    let foo_from_builder: Foo = FooBuilder::new().name("Y").build();
```
Comment (pickfire) [inline patterns/builder.md:54]: Do we need this type signature, isn't it inferred?

File: patterns/builder.md [hunk 9]
```diff
@@ -2,46 +2,58 @@
...
         // Set the minimally required fields of Foo.
+        FooBuilder {
+            bar: "X".to_string(),
+        }
     }
 
-    fn named(mut self, name: &str) -> FooBuilder {
+    pub fn name(mut self, bar: &str) -> FooBuilder {
         // Set the name on the builder itself, and return the builder by value.
+        self.bar = bar.to_string();
```
Comment (pickfire) [inline patterns/builder.md:37]: Is this a good pattern to clone within the function when users can provide `String`? I am not quite sure how this is done in the wild.

File: patterns/builder.md (hunk 8 above)
Comment (TGotwig) [reply patterns/builder.md:48]: It's not needed but best practice to do so, I think 🤔

Comment (pickfire) [reply patterns/builder.md]: I am not sure if it is best practice but for docs examples (including rust), most types are explicitly state to let others to understand easily but I think this example should be way to simple, if they get to this page they should already be able to see through this simple example.

**Review by pickfire — APPROVED** [review]
Looks good to me. I think it would be good to have one of the maintainer to check this though.

**Comment by simonsan** [comment]
@TGotwig Hey, thank you for the PR. Could you please rebase on `master` so the CI is checking this through while we get to reviewing this PR? Thank you ;-)

**Comment by TGotwig** [comment]
@simonsan Sure, will do! 😃

**Comment by simonsan** [comment]
> @simonsan Sure, will do! 😃

Sorry, haven't seen that there is no addition to summary or anything needed. So I could do it myself. ;-) It's fine then.

**Comment by simonsan** [comment]
@MarcoIeni You also want to look over again?

**Review by MarcoIeni — APPROVED** [review]

**Comment by simonsan** [comment]
Thanks! :-)

**Comment by TGotwig** [comment]
Cool! 😊🚀

**Pull request merged**